File: data-management/viewer/backend/src/api/models/annotations.py
"""
Annotation Pydantic models for robotic episode annotation system.

These models match the TypeScript type definitions and PRD schema specifications
for task completeness, trajectory quality, data quality, and anomaly annotations.
"""

from datetime import datetime
from enum import Enum, StrEnum
from typing import Annotated, ClassVar

from pydantic import Field

from ..validation import SanitizedModel

# ============================================================================
# Task Completeness Types
# ============================================================================


class TaskCompletenessRating(StrEnum):
    """Task completion status rating."""

    SUCCESS = "success"
    PARTIAL = "partial"
    FAILURE = "failure"
    UNKNOWN = "unknown"


class ConfidenceLevel(int, Enum):
    """Annotator confidence level (1-5 scale)."""

    ONE = 1
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5


class TaskCompletenessAnnotation(SanitizedModel):
    """Task completeness annotation with rating and optional details."""

    rating: TaskCompletenessRating
    confidence: ConfidenceLevel
    completion_percentage: int | None = Field(None, ge=0, le=100)
    failure_reason: str | None = None
    subtask_reached: str | None = None

    model_config: ClassVar = {"use_enum_values": True}


# ============================================================================
# Trajectory Quality Types
# ============================================================================


class QualityScore(int, Enum):
    """Quality score on a 1-5 scale."""

    ONE = 1
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5


class TrajectoryFlag(StrEnum):
    """Flags indicating specific trajectory issues."""

    JITTERY = "jittery"
    INEFFICIENT_PATH = "inefficient-path"
    NEAR_COLLISION = "near-collision"
    OVER_EXTENSION = "over-extension"
    UNDER_REACHING = "under-reaching"
    HESITATION = "hesitation"
    CORRECTION_HEAVY = "correction-heavy"


class TrajectoryQualityMetrics(SanitizedModel):
    """Individual trajectory quality metrics."""

    smoothness: QualityScore
    efficiency: QualityScore
    safety: QualityScore
    precision: QualityScore

    model_config: ClassVar = {"use_enum_values": True}


class TrajectoryQualityAnnotation(SanitizedModel):
    """Complete trajectory quality annotation."""

    overall_score: QualityScore
    metrics: TrajectoryQualityMetrics
    flags: list[TrajectoryFlag] = Field(default_factory=list)

    model_config: ClassVar = {"use_enum_values": True}


# ============================================================================
# Data Quality Types
# ============================================================================


class DataQualityLevel(StrEnum):
    """Overall data quality level."""

    GOOD = "good"
    ACCEPTABLE = "acceptable"
    POOR = "poor"
    UNUSABLE = "unusable"


class DataQualityIssueType(StrEnum):
    """Types of data quality issues."""

    FRAME_DROP = "frame-drop"
    SYNC_ISSUE = "sync-issue"
    OCCLUSION = "occlusion"
    LIGHTING_ISSUE = "lighting-issue"
    SENSOR_NOISE = "sensor-noise"
    CALIBRATION_DRIFT = "calibration-drift"
    ENCODING_ARTIFACT = "encoding-artifact"
    MISSING_DATA = "missing-data"


class IssueSeverity(StrEnum):
    """Severity of a data quality issue."""

    MINOR = "minor"
    MAJOR = "major"
    CRITICAL = "critical"


class DataQualityIssue(SanitizedModel):
    """Individual data quality issue."""

    type: DataQualityIssueType
    severity: IssueSeverity
    affected_frames: tuple[int, int] | None = None
    affected_streams: list[str] | None = None
    notes: str | None = None

    model_config: ClassVar = {"use_enum_values": True}


class DataQualityAnnotation(SanitizedModel):
    """Complete data quality annotation."""

    overall_quality: DataQualityLevel
    issues: list[DataQualityIssue] = Field(default_factory=list)

    model_config: ClassVar = {"use_enum_values": True}


# ============================================================================
# Anomaly Types
# ============================================================================


class AnomalySeverity(StrEnum):
    """Severity of an anomaly."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class AnomalyType(StrEnum):
    """Types of anomalies that can be detected."""

    UNEXPECTED_STOP = "unexpected-stop"
    TRAJECTORY_DEVIATION = "trajectory-deviation"
    FORCE_SPIKE = "force-spike"
    VELOCITY_SPIKE = "velocity-spike"
    OBJECT_SLIP = "object-slip"
    GRIPPER_FAILURE = "gripper-failure"
    COLLISION = "collision"
    OTHER = "other"


class Anomaly(SanitizedModel):
    """Individual anomaly marker."""

    id: str
    type: AnomalyType
    severity: AnomalySeverity
    frame_range: tuple[int, int]
    timestamp: tuple[float, float]
    description: str
    auto_detected: bool = False
    verified: bool = False

    model_config: ClassVar = {"use_enum_values": True}


class AnomalyAnnotation(SanitizedModel):
    """Container for anomaly annotations."""

    anomalies: list[Anomaly] = Field(default_factory=list)


# ============================================================================
# Language Instruction Types (VLA)
# ============================================================================


class InstructionSource(StrEnum):
    """Source of the language instruction annotation."""

    HUMAN = "human"
    TEMPLATE = "template"
    LLM_GENERATED = "llm-generated"
    RETROACTIVE = "retroactive"


class LanguageInstructionAnnotation(SanitizedModel):
    """Natural language instruction for VLA-conditioned training.

    Stores the primary task instruction plus optional paraphrases and
    subtask decomposition used for data augmentation and hierarchical
    policy conditioning.
    """

    instruction: str = Field(min_length=1, max_length=1000)
    source: InstructionSource
    language: str = Field(default="en", max_length=10)
    paraphrases: list[Annotated[str, Field(max_length=1000)]] = Field(default_factory=list, max_length=50)
    subtask_instructions: list[Annotated[str, Field(max_length=1000)]] = Field(default_factory=list, max_length=100)

    model_config: ClassVar = {"use_enum_values": True}


# ============================================================================
# Object Detection Annotation Types
# ============================================================================


class ObjectDetectionBox(SanitizedModel):
    """Single saved object detection with label and bounding box."""

    label: str = Field(min_length=1, max_length=100)
    confidence: float = Field(ge=0.0, le=1.0)
    bbox: tuple[float, float, float, float] = Field(description="Pixel bounding box (x1, y1, x2, y2)")


class ObjectDetectionAnnotation(SanitizedModel):
    """Open-vocabulary object detections saved for a single reference frame."""

    frame_index: int = Field(ge=0)
    camera: str = Field(default="il-camera", max_length=64)
    queried_labels: list[Annotated[str, Field(max_length=100)]] = Field(default_factory=list, max_length=64)
    detections: list[ObjectDetectionBox] = Field(default_factory=list, max_length=200)
    model: str | None = Field(default=None, max_length=64)


# ============================================================================
# Combined Episode Annotation Types
# ============================================================================


class EpisodeAnnotation(SanitizedModel):
    """Complete annotation for a single episode by one annotator."""

    annotator_id: str
    timestamp: datetime
    task_completeness: TaskCompletenessAnnotation
    trajectory_quality: TrajectoryQualityAnnotation
    data_quality: DataQualityAnnotation
    anomalies: AnomalyAnnotation
    language_instruction: LanguageInstructionAnnotation | None = None
    object_detections: list[ObjectDetectionAnnotation] = Field(default_factory=list, max_length=32)
    notes: str | None = None


class EpisodeConsensus(SanitizedModel):
    """Consensus annotation derived from multiple annotators."""

    task_completeness: TaskCompletenessRating
    trajectory_score: float = Field(ge=1.0, le=5.0)
    data_quality: DataQualityLevel
    agreement_score: float = Field(ge=0.0, le=1.0)

    model_config: ClassVar = {"use_enum_values": True}


class EpisodeAnnotationFile(SanitizedModel):
    """Complete annotation file for an episode."""

    schema_version: str = "1.0.0"
    episode_index: int = Field(ge=0)
    dataset_id: str
    annotations: list[EpisodeAnnotation] = Field(default_factory=list)
    consensus: EpisodeConsensus | None = None


# ============================================================================
# Auto-Analysis Types
# ============================================================================


class ComputedQualityMetrics(SanitizedModel):
    """Computed trajectory quality metrics from auto-analysis."""

    smoothness_score: float = Field(ge=0.0, le=1.0)
    efficiency_score: float = Field(ge=0.0, le=1.0)
    jitter_metric: float = Field(ge=0.0)
    hesitation_count: int = Field(ge=0)
    correction_count: int = Field(ge=0)


class AutoQualityAnalysis(SanitizedModel):
    """Auto-analysis result for an episode."""

    episode_index: int = Field(ge=0)
    computed: ComputedQualityMetrics
    suggested_rating: int = Field(ge=1, le=5)
    confidence: float = Field(ge=0.0, le=1.0)
    flags: list[TrajectoryFlag] = Field(default_factory=list)

    model_config: ClassVar = {"use_enum_values": True}


# ============================================================================
# Annotation Summary Types
# ============================================================================


class AnnotationSummary(SanitizedModel):
    """Aggregated annotation metrics for a dataset."""

    dataset_id: str
    total_episodes: int = Field(ge=0)
    annotated_episodes: int = Field(ge=0)
    task_completeness_distribution: dict[str, int] = Field(default_factory=dict)
    quality_score_distribution: dict[int, int] = Field(default_factory=dict)
    anomaly_type_counts: dict[str, int] = Field(default_factory=dict)

File: data-management/viewer/frontend/src/types/annotations.ts
/**
 * Annotation type definitions for robotic episode annotation system.
 * These types match the PRD schema definitions for task completeness,
 * trajectory quality, data quality, and anomaly annotations.
 */

// ============================================================================
// Task Completeness Types
// ============================================================================

/** Task completion status rating */
export type TaskCompletenessRating = 'success' | 'partial' | 'failure' | 'unknown'

/** Annotator confidence level (1-5 scale) */
export type ConfidenceLevel = 1 | 2 | 3 | 4 | 5

/** Task completeness annotation with rating and optional details */
export interface TaskCompletenessAnnotation {
  /** Overall task completion rating */
  rating: TaskCompletenessRating
  /** Annotator confidence in the rating */
  confidence: ConfidenceLevel
  /** Percentage of task completed (0-100), used for partial ratings */
  completionPercentage?: number
  /** Reason for failure, used when rating is 'failure' */
  failureReason?: string
  /** ID of the last completed subtask, used for partial ratings */
  subtaskReached?: string
}

// ============================================================================
// Trajectory Quality Types
// ============================================================================

/** Quality score on a 1-5 scale */
export type QualityScore = 1 | 2 | 3 | 4 | 5

/** Flags indicating specific trajectory issues */
export type TrajectoryFlag =
  | 'jittery'
  | 'inefficient-path'
  | 'near-collision'
  | 'over-extension'
  | 'under-reaching'
  | 'hesitation'
  | 'correction-heavy'

/** Individual trajectory quality metrics */
export interface TrajectoryQualityMetrics {
  /** Jitter-free motion score */
  smoothness: QualityScore
  /** Path optimality score */
  efficiency: QualityScore
  /** Collision avoidance score */
  safety: QualityScore
  /** End-effector accuracy score */
  precision: QualityScore
}

/** Complete trajectory quality annotation */
export interface TrajectoryQualityAnnotation {
  /** Overall trajectory quality score */
  overallScore: QualityScore
  /** Individual metric scores */
  metrics: TrajectoryQualityMetrics
  /** Flags for specific issues observed */
  flags: TrajectoryFlag[]
}

// ============================================================================
// Data Quality Types
// ============================================================================

/** Overall data quality level */
export type DataQualityLevel = 'good' | 'acceptable' | 'poor' | 'unusable'

/** Types of data quality issues */
export type DataQualityIssueType =
  | 'frame-drop'
  | 'sync-issue'
  | 'occlusion'
  | 'lighting-issue'
  | 'sensor-noise'
  | 'calibration-drift'
  | 'encoding-artifact'
  | 'missing-data'

/** Severity of a data quality issue */
export type IssueSeverity = 'minor' | 'major' | 'critical'

/** Individual data quality issue */
export interface DataQualityIssue {
  /** Type of issue */
  type: DataQualityIssueType
  /** Severity of the issue */
  severity: IssueSeverity
  /** Frame range affected [start, end] */
  affectedFrames?: [number, number]
  /** Camera names or sensor IDs affected */
  affectedStreams?: string[]
  /** Additional notes about the issue */
  notes?: string
}

/** Complete data quality annotation */
export interface DataQualityAnnotation {
  /** Overall data quality assessment */
  overallQuality: DataQualityLevel
  /** List of specific issues found */
  issues: DataQualityIssue[]
}

// ============================================================================
// Anomaly Types
// ============================================================================

/** Severity of an anomaly */
export type AnomalySeverity = 'low' | 'medium' | 'high'

/** Types of anomalies that can be detected */
export type AnomalyType =
  | 'unexpected-stop'
  | 'trajectory-deviation'
  | 'force-spike'
  | 'velocity-spike'
  | 'object-slip'
  | 'gripper-failure'
  | 'collision'
  | 'other'

/** Individual anomaly marker */
export interface Anomaly {
  /** Unique identifier for this anomaly */
  id: string
  /** Type of anomaly */
  type: AnomalyType
  /** Severity level */
  severity: AnomalySeverity
  /** Frame range [start, end] */
  frameRange: [number, number]
  /** Timestamp range in seconds [start, end] */
  timestamp: [number, number]
  /** Human-readable description */
  description: string
  /** Whether this was auto-detected by the system */
  autoDetected: boolean
  /** Whether a human has verified this anomaly */
  verified: boolean
}

/** Container for anomaly annotations */
export interface AnomalyAnnotation {
  /** List of anomalies in this episode */
  anomalies: Anomaly[]
}

// ============================================================================
// Language Instruction Types (VLA)
// ============================================================================

/** Source of the language instruction */
export type InstructionSource = 'human' | 'template' | 'llm-generated' | 'retroactive'

/** Language instruction annotation for VLA-conditioned training */
export interface LanguageInstructionAnnotation {
  /** Primary task instruction */
  instruction: string
  /** How this instruction was produced */
  source: InstructionSource
  /** ISO 639-1 language code */
  language: string
  /** Alternative phrasings for data augmentation */
  paraphrases: string[]
  /** Ordered subtask decomposition */
  subtaskInstructions: string[]
}

// ============================================================================
// Object Detection Annotation Types
// ============================================================================

/** Single saved object detection with label and bounding box. */
export interface ObjectDetectionBox {
  /** Label/class name as queried by the user (or COCO class for closed-vocabulary). */
  label: string
  /** Detection confidence in [0, 1]. */
  confidence: number
  /** Pixel bounding box (x1, y1, x2, y2) in source-image coordinates. */
  bbox: [number, number, number, number]
}

/** Open-vocabulary object detections saved for a single reference frame. */
export interface ObjectDetectionAnnotation {
  /** Frame index the detections were computed on. */
  frameIndex: number
  /** Camera stream the frame was sourced from. */
  camera: string
  /** User-supplied labels passed to the open-vocabulary detector (empty for closed-vocabulary). */
  queriedLabels: string[]
  /** Detection results, one per box. */
  detections: ObjectDetectionBox[]
  /** Model variant used to produce the detections. */
  model?: string
}

// ============================================================================
// Combined Episode Annotation Types
// ============================================================================

/** Complete annotation for a single episode by one annotator */
export interface EpisodeAnnotation {
  /** Unique identifier for the annotator */
  annotatorId: string
  /** ISO 8601 timestamp when annotation was created/updated */
  timestamp: string
  /** Task completeness annotation */
  taskCompleteness: TaskCompletenessAnnotation
  /** Trajectory quality annotation */
  trajectoryQuality: TrajectoryQualityAnnotation
  /** Data quality annotation */
  dataQuality: DataQualityAnnotation
  /** Anomaly annotations */
  anomalies: AnomalyAnnotation
  /** Language instruction for VLA training */
  languageInstruction?: LanguageInstructionAnnotation
  /** Saved open-vocabulary object detections per reference frame */
  objectDetections?: ObjectDetectionAnnotation[]
  /** Free-form notes about the episode */
  notes?: string
}

/** Consensus annotation derived from multiple annotators */
export interface EpisodeConsensus {
  /** Consensus task completeness rating */
  taskCompleteness: TaskCompletenessRating
  /** Average trajectory quality score */
  trajectoryScore: number
  /** Consensus data quality level */
  dataQuality: DataQualityLevel
  /** Inter-annotator agreement score (0-1) */
  agreementScore: number
}

/** Complete annotation file for an episode */
export interface EpisodeAnnotationFile {
  /** Schema version for this file format */
  schemaVersion: string
  /** Episode index within the dataset */
  episodeIndex: number
  /** Dataset identifier */
  datasetId: string
  /** List of annotations from different annotators */
  annotations: EpisodeAnnotation[]
  /** Computed consensus (if multiple annotators) */
  consensus?: EpisodeConsensus
}

// ============================================================================
// Default/Initial Values
// ============================================================================

/** Create a default task completeness annotation */
export function createDefaultTaskCompleteness(): TaskCompletenessAnnotation {
  return {
    rating: 'unknown',
    confidence: 3,
  }
}

/** Create a default trajectory quality annotation */
export function createDefaultTrajectoryQuality(): TrajectoryQualityAnnotation {
  return {
    overallScore: 3,
    metrics: {
      smoothness: 3,
      efficiency: 3,
      safety: 3,
      precision: 3,
    },
    flags: [],
  }
}

/** Create a default data quality annotation */
export function createDefaultDataQuality(): DataQualityAnnotation {
  return {
    overallQuality: 'good',
    issues: [],
  }
}

/** Create a default anomaly annotation */
export function createDefaultAnomalyAnnotation(): AnomalyAnnotation {
  return {
    anomalies: [],
  }
}

/** Create a default language instruction annotation */
export function createDefaultLanguageInstruction(): LanguageInstructionAnnotation {
  return {
    instruction: '',
    source: 'human',
    language: 'en',
    paraphrases: [],
    subtaskInstructions: [],
  }
}

/** Create a complete default episode annotation */
export function createDefaultEpisodeAnnotation(annotatorId: string): EpisodeAnnotation {
  return {
    annotatorId,
    timestamp: new Date().toISOString(),
    taskCompleteness: createDefaultTaskCompleteness(),
    trajectoryQuality: createDefaultTrajectoryQuality(),
    dataQuality: createDefaultDataQuality(),
    anomalies: createDefaultAnomalyAnnotation(),
  }
}
